core/src/rpc/aggregator.rs

```
@@ -583,6 +616,61 @@ impl Aggregator {
         Ok(move_txhandler.promote()?)
     }
 
+    async fn verify_and_save_emergency_stop_sigs(
+        &self,
+        emergency_stop_sigs: Vec<Vec<u8>>,
+        emergency_stop_agg_nonce: MusigAggNonce,
+        deposit_params: DepositParams,
+    ) -> Result<(), BridgeError> {
+        let mut deposit_data: crate::builder::transaction::DepositData =
+            deposit_params.try_into()?;
+        let musig_partial_sigs = parser::verifier::parse_partial_sigs(emergency_stop_sigs)?;
+
+        // create move tx and calculate sighash
+        let move_txhandler =
+            create_move_to_vault_txhandler(&mut deposit_data, self.config.protocol_paramset())?;
+
+        let mut emergency_stop_txhandler = create_emergency_stop_txhandler(
+            &mut deposit_data,
+            &move_txhandler,
+            self.config.protocol_paramset(),
+        )?;
+
+        let sighash = emergency_stop_txhandler.calculate_script_spend_sighash_indexed(
+            0,
+            0,
+            bitcoin::TapSighashType::SinglePlusAnyoneCanPay,
+        )?;
+
+        let verifiers_public_keys = deposit_data.get_verifiers();
+
+        let final_sig = crate::musig2::aggregate_partial_signatures(
+            &verifiers_public_keys,
+            None,
+            emergency_stop_agg_nonce,
+            &musig_partial_sigs,
+            Message::from_digest(sighash.to_byte_array()),
+        )
+        .map_err(|x| {
+            BridgeError::Error(format!(
+                "Aggregating Emergency Stop signatures failed {}",
+                x
```

this discards the error chain information, prefer `wrap_err("Failed to aggregate emergency stop signatures")` (you may need to import eyre::Context)

---

6f3eb32